vogue/server/utils/bioinfo/micro.py

```diff
@@ -0,0 +1,370 @@
+            }
+        }
+    }, {
+        '$group': {
+            '_id': {
+                'month': '$month'
+            },
+            'ST': {
+                '$push': '$ST'
+            }
+        }
+    }]
+
+    intermediate = {}
+    aggregate_result = adapter.bioinfo_samples_aggregate(pipe)
+    for result in aggregate_result:
+        untyped = 0
+        typed = 0
+        untyped_lower_threshold = -9
+        untyped_upper_threshold = -1
+        for sequence_type in result['ST']:
+            try:
+                sequence_type = int(sequence_type)
+                if untyped_lower_threshold < sequence_type < untyped_upper_threshold:
+                    untyped += 1
+                else:
+                    typed += 1
+            except:
+                pass
+        if typed + untyped:
+            fraction = untyped / (typed + untyped)
+        else:
+            fraction = None
+        intermediate[result['_id']['month']] = fraction
+
+    data = []
+    labels = []
+    for m in MONTHS:
+        value = intermediate.get(m[0], None)
+        data.append(value)
+        labels.append(m[1])
+
+    plot_data = {'data': data, 'labels': labels}
+    return plot_data
+
+
+def microsalt_get_st_time(adapter, year: int) -> dict:
+    pipe = [{
+        '$match': {
+            'microsalt': {
+                '$exists': 'True'
+            }
+        }
+    }, {
+        '$lookup': {
+            'from': 'sample',
+            'localField': '_id',
+            'foreignField': '_id',
+            'as': 'sample_info'
+        }
+    }, {
+        '$unwind': {
+            'path': '$sample_info'
+        }
+    }, {
+        '$match': {
+            'sample_info.received_date': {
+                '$exists': 'True'
+            }
+        }
+    }, {
+        '$project': {
+            'month': {
+                '$month': '$sample_info.received_date'
+            },
+            'strain': '$sample_info.strain',
+            'year': {
+                '$year': '$sample_info.received_date'
+            },
+            'sequence_type': '$microsalt.blast_pubmlst.sequence_type'
+        }
+    }, {
+        '$match': {
+            'year': {
+                '$eq': int(year)
+            },
+            'sequence_type': {
+                '$exists': 'True'
+            },
+            'strain': {
+                '$exists': 'True'
+            }
+        }
+    }, {
+        '$group': {
+            '_id': {
+                'month': '$month',
+                'strain': '$strain',
+                'sequence_type': '$sequence_type'
+            },
+            'number': {
+                '$sum': 1
+            }
+        }
+    }]
+    final_results = {}
+    aggregate_result = list(adapter.bioinfo_samples_aggregate(pipe))
+    for result in aggregate_result:
+        st = result['_id']['sequence_type']
+        strain = result['_id']['strain']
+        if final_results.get(strain):
+            final_results[strain][st] = [None] * 12
```

Comment:
I know it is obvious a year has 12 month...

```suggestion
            final_results[strain][st] = [None] * len(MONTHS)
```